### ISSUE TITLE
The translation function changes during rendering

### ISSUE THREAD

**Issue by iksent** [vinissimus/next-translate#447]
**Problem:** The translation function changes during rendering
I experiencing this issue with the next code:

```
import React, { useEffect, useState } from 'react'
import useTranslation from 'next-translate/useTranslation'
import Button from '../Button/Button'

const FormErrorHandler = () => {
  const [state, setState] = useState(false)
  const { t } = useTranslation('validation')

  useEffect(() => {
    console.log('T WAS CHANGED!', state)
  }, [t])

  return <Button title={t('email')}  onClick={() => setState((p) => !p)} />
}

export default FormErrorHandler

```

Console output:
```
T WAS CHANGED! false
T WAS CHANGED! true
T WAS CHANGED! false
...
```

I think this is a big problem: I am passing `t` function to any hook, that need to be recreated after language changed.

There was no issues using [next-i18next](https://github.com/isaachinman/next-i18next) with it: Translate function was changed only when language changed.

Currently I can avoid this problem with `useMemo` hook like:

```
import React, { useEffect, useMemo, useState } from 'react'
import useTranslation from 'next-translate/useTranslation'
import Button from '../Button/Button'

const _useTranslation = () => {
  const { t, lang } = useTranslation('validation')
  const T = useMemo(() => t, [lang])
  return { T, lang }
}

const FormErrorHandler = () => {
  const [state, setState] = useState(false)
  const { T } = _useTranslation()

  useEffect(() => {
    console.log('T WAS CHANGED!', state)
  }, [T])

  return <Button title={T('email')} onClick={() => setState((p) => !p)} />
}

export default FormErrorHandler
```

... due to I have a wrapper function for your library.
But this could be a really big problem for all who hasn't.

**Comment by aralroca** [comment]
But this is not really a problem... you put it as a dependency of the hook but in reality it would only make sense to put the `lang` as the dependency, right?

```js
import React, { useEffect, useState } from 'react'
import useTranslation from 'next-translate/useTranslation'
import Button from '../Button/Button'

const FormErrorHandler = () => {
  const [state, setState] = useState(false)
  const { t, lang } = useTranslation('validation')

  useEffect(() => {
    // ...
  }, [lang]) // doesn't make sense to put the `t` function as dependency

  return <Button title={t('email')}  onClick={() => setState((p) => !p)} />
}

export default FormErrorHandler
```

I also recommend using the `t` function only in the template to avoid problems:

```js
function Example() {
  const { t } = useTranslation('validation')
  const [state, set] = useReducer(reducer, initialArg, init);

  useEffect(() => {
    set('translationKey', 'example-of-translation')
  }, 
  []) // No dependencies
  
  return <div>{t(state.translationKey}</div>
}
```

**Comment by iksent** [comment]
1. Yes, I can put `lang` as dependency, but after migration from [i18next](https://www.i18next.com/), I need to change it everywhere in the project (this is probably 100+ places)
2. Don't you think this library should work as much similar as other popular i18n libraries? I don't really know, if every JS library works similar, but I think `next-translate` is very similar to [i18next](https://www.i18next.com/) now.
3. "I also recommend using the t function only in the template to avoid problems:"
I was talking about dependecy for `useCallback` and `useMemo`, which return ReactNode.

**Comment by RandomDude123** [comment]
Ran into the same issue. Fixed it by removing `t` from dependency array and  adding `// eslint-disable-next-line react-hooks/exhaustive-deps`. I do agree with @iksent that this can lead to serious issues for people not realising that `t` changes on render.

**Comment by aralroca** [comment]
@iksent @RandomDude123 I did a PR to fix this inside the `useTranslation` hook, this way you do not need a workaround.

**Comment by aralroca** [comment]
@iksent @RandomDude123  already pre-released on [1.0.6-canary.3](https://github.com/vinissimus/next-translate/releases/tag/1.0.6-canary.3) 😊  If you can confirm that it works, that would be great

**Comment by aralroca** [comment]
I think we can revert the PR, because it also does not solve the problem it was trying to solve and introduces this other one: https://github.com/vinissimus/next-translate/issues/604

**Comment by timohofmeijer** [comment]
@aralroca Memoizing `t` does prevent translations from being removed before any page/route transition animation. As of now, people should be aware they’ll have to memoize `t` themselves when using page transitions, wouldn’t it be more appropriate to solve this within the library.

**Comment by timohofmeijer** [comment]
For peoples needing a fix for this, perhaps this can save you some time.

Create a custom hook which returns a memoized copy of `t`.

```ts
// /src/hooks/use-memo-translation.ts

import { useMemo } from 'react'
import { Translate } from 'next-translate'
import useTranslation from 'next-translate/useTranslation'

/**
  In the current version of next-translate `t` will change before
  any page-transition-animation starts. Memoizing `t` will keep it
  around while animating.
*/
const _useTranslation = (key?: string): { t: Translate; lang: string } => {
  const { t: _t, lang } = useTranslation(key)
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const t = useMemo(() => _t, [lang]) // `t` should NOT be a hook dependency here.
  return { t, lang }
}

//  Default export to make reverting to non wrapped version more seamless
export default _useTranslation
```

Next making sure the `baseUrl` in `tsconfig.json` is set ([docs](https://nextjs.org/docs/advanced-features/module-path-aliases)) will allow you te replace all `next-translate/useTranslation` imports with `hooks/use-memo-translation` and makes you easily revert to the original hook.

**Comment by benvium** [comment]
@timohofmeijer Thanks! I ran into unexpected useEffect infinite loops due to this issue too. It really kills any useEffect or useMemo usage unfortunately.

I tried your hook and the tsconfig baseUrl suggestion you made:
```
"baseUrl": ".",
    "paths": {
      "next-translate/useTranslation":["src/feature/i18n/useMemoTranslation.ts"]
    }
```

.. but got a different infinite loop - because `import useTranslation from 'next-translate/useTranslation'` inside useMemoTranslation was being re-written to itself.

I suppose it'd work if use-memo-translation was a separate library?

**Comment by timohofmeijer** [comment]
> @timohofmeijer Thanks! I ran into unexpected useEffect infinite loops due to this issue too. It really kills any useEffect or useMemo usage unfortunately.
> 
> I tried your hook and the tsconfig baseUrl suggestion you made:
> 
> ```
> "baseUrl": ".",
>     "paths": {
>       "next-translate/useTranslation":["src/feature/i18n/useMemoTranslation.ts"]
>     }
> ```
> 
> .. but got a different infinite loop - because `import useTranslation from 'next-translate/useTranslation'` inside useMemoTranslation was being re-written to itself.
> 
> I suppose it'd work if use-memo-translation was a separate library?

I used `baseUrl` to get rid of relative import paths to make it easy to replace back all imports in my codebase. I simply set `"baseUrl": "./src",` so I could import like this `import useTranslate from 'hooks/use-memo-translation'`. Making it easy to replace could have been achieved by a simple regex search/replace as well though. It was not actually part of the fix, sorry for putting you on the wrong track there.

**Comment by benvium** [comment]
Thanks @timohofmeijer! I ended up manually using `useMemoTranslation` instead - likely clearer for future maintainers of this code too, rather than tsconfig 'magic'. All works well so far.

**Comment by Ainias** [comment]
Is there a fix for this? in 1.5.0 it is still a problem where I think it could be a really simple fix inside the library. 
Especially since `eslint-disable-next-line react-hooks/exhaustive-deps` wants to add the `t` function to every dependency-array
